### PR TITLE
Merge api-devfile and api-workspace together

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
... and put the infrastructure-dependent code behind injectable interfaces.

I had to update ASM (to the last version 7.1) ~~and explicitly set dependency on `javax.json:json-api` both of~~ which will probably require to go through the CQ process.

### What issues does this PR fix or reference?
#13315

### CODE REVIEW

**Comment by metlos** [comment]
I created an associated PR for the library updates in `che-parent`: https://github.com/eclipse/che-parent/pull/117

**Comment by metlos** [comment]
ci-build

**Comment by metlos** [comment]
Ah, this won't pass the build check until https://github.com/eclipse/che-parent/pull/117 is merged.

File: wsmaster/che-core-api-workspace/pom.xml
Comment (mshaposhnik) [inline wsmaster/che-core-api-workspace/pom.xml]: Remove version

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/convert/component/plugin/PluginComponentToWorkspaceApplier.java
Comment (mshaposhnik) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/convert/component/plugin/PluginComponentToWorkspaceApplier.java]: Looks like your editor hates static imports

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java
Comment (amisevsk) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java]: If this test case no longer has anything to do with OpenShift the name should be changed.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java
Comment (amisevsk) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java]:
```suggestion
   * identify endpoint as discoverable (i.e. it is accessible by its name from workspace's
```

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (amisevsk) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: Is this copied from the previous `DevfileModule` or different file? If the former, it'd be great to keep the git history with a move, but it's altogether not that important.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplierTest.java]: I think it shold be reverted since it duplicates another test fully now

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java]: \+

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesEnvironmentProvisioner.java
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesEnvironmentProvisioner.java]: \+

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesEnvironmentProvisioner.java
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesEnvironmentProvisioner.java:59]: \+

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentIntegrityValidatorTest.java
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentIntegrityValidatorTest.java]: \+

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java
Comment (mshaposhnik) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java]: \+

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java]: Shared classes are in `org.eclipse.che.api.workspace.shared.dto.devfile` package.

Stack related classes are is `org.eclipse.che.api.workspace.server.stack` package.

Maybe it would be better to name devfile server package as ` org.eclipse.che.api.workspace.server.devfile`?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: Looks like we do not need `DevfileToWorkspaceConfigConverter` anymore after merging these two modules since it should be possible to use `DevfileConverter` directly in `WorkspaceRuntimes`. WDYT?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java]: Also, devfile related impls are in `org.eclipse.che.api.workspace.server.model.impl.devfile`

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: I'm not sure that we really need `DevfileBindings` but if we have it, why do not reuse it for `ComponentProvisioner` and `ComponentToWorkspaceApplier` in this class?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileBindings.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileBindings.java]: Could you elaborate where Map<String, ComponentProvisioner> is used?
As far as I see set binder should be used here https://github.com/eclipse/che/blob/8380b7ce55853e4fb9da12685483aa941fb8641d/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/convert/DevfileConverter.java#L65

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java]: Since Kubernetes component related classes are moved to k8s-infra module, I think it makes sense to move related constants as well.
It's not a requirement but just a recommendation to consider doing it since I realize that it may require a lot of additional work that is not so needed.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesDevfileBindings.java
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesDevfileBindings.java]: Bidings` -> `Bindings`?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/DockerimageComponentToWorkspaceApplier.java
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/DockerimageComponentToWorkspaceApplier.java:12]: From the first look, it's not clear why Dockerimage component related classes are in k8s module.
Maybe it makes sense to add into java docs that this class converts dockerimage component to k8s infra understood format?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java]: Please consider removing this unused method.

File: wsmaster/che-core-api-devfile/pom.xml
Comment (sleshchenko) [inline wsmaster/che-core-api-devfile/pom.xml:22]: Looks like they're still 3 places where this artifact is used. Build works locally only because of existence outdated artifact in local maven repo. Please make sure that it is not used anymore otherwise build after next release will fail
Screenshot_20190528_144548

Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/pom.xml:22]: Also, please update CODEOWNERS

https://github.com/eclipse/che/blob/3d3f4a463cc02ee6af5cacda93df6cf48e3d0413/.github/CODEOWNERS#L51

**Review by sleshchenko — COMMENTED** [review]
Generally looks good but it's not ready to merge yet.

Please let me know when all the comments will be addressed.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesEnvironmentProvisioner.java
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesEnvironmentProvisioner.java]: IMHO, "either foo or bar or baz" is still a correct English expression but never mind. "any of foo or bar or baz" works as well.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentIntegrityValidatorTest.java
Comment (metlos) [reply infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentIntegrityValidatorTest.java:36]: Yes, :+1:. Tests present at the moment were just moved from the old test cases.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: It was definitely moved. Not sure why git considers it remove and add.

Comment (amisevsk) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: Yeah sometimes it breaks for me (especially if the move is done via IDE).

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileManager.java
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileManager.java]: I do :) The `@see` references a private method in `WorkspaceManager` which IJ complains about for me.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: Yeah, I first added `KubernetesDevfileBindings` that contains a non-trivial setup of the binding which I thought warranted making sure that we don't mess it up on the places it is needed (https://github.com/eclipse/che/blob/8f88534bc253087f19f0085686ad5148ed1dc481/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesDevfileBindings.java#L78).

I liked the way how it read in the module definitions so I added also the trivial bindings in the `DevfileBindings`.

If you believe we should be explicit about the binding setup in the modules, I can remove these classes but I believe it makes the binding definitions ever so slightly less error prone.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileBindings.java
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileBindings.java]: Huh, good catch!

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileModule.java
Comment (sleshchenko) [inline wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/devfile/DevfileModule.java]: It's not related to your changes, but this binding looks unneeded.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileManager.java
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileManager.java]: Then maybe it would be better to fix an issue instead of suppress warning? 
The referenced private method tells nothing about a format, it contains code that parse key, but not java doc at all. From Java Doc perspective it would be better to reference 
```
public WorkspaceImpl getWorkspace(String key)
```
where key rules are described.
WDYT?

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/Constants.java]: > Since Kubernetes component related classes are moved to k8s-infra module, I think it makes sense to move related constants as well.
It's not a requirement but just a recommendation to consider doing it since I realize that it may require a lot of additional work that is not so needed.

@metlos Could you please leave your opinion about it? The comment is resolved but plan about this topic is not clear.

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: DevfileToWorkspaceConfigConverter is removed now

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java
Comment (sleshchenko) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileModule.java]: I'm OK with leaving `DevfileBindings` class

File: wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileManager.java
Comment (metlos) [reply wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/devfile/server/DevfileManager.java]: :+1: I didn't change the method, so I didn't want to change the javadoc either and frankly didn't look into the method itself. Will change it.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java:65]: `component's referenced content` -> `components referenced content`?

**Review by sleshchenko — APPROVED** [review]
LGTM :+1: 
Can't wait to see it merged :smile:

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java
Comment (metlos) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java:65]: component's referenced content ~ referenced content of the component, so it is fine as is imho

**Review by mshaposhnik — APPROVED** [review]

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentValidator.java:65]: sorry, you're right.

**Review by skabashnyuk — APPROVED** [review]

**Comment by mshaposhnik** [comment]
ci-build

**Comment by mshaposhnik** [comment]
ci-build

**Comment by mshaposhnik** [comment]
ci-build

**Comment by mshaposhnik** [comment]
ci-test

**Pull request merged**
